File: Cube/Geometry/Vector.cs
using System;

namespace Cube.Geometry
{
    class Vector
    {
        public double X { get; private set; }
        public double Y { get; private set; }
        public double Z { get; private set; }
        public Vector(double x, double y, double z)
        {
            X = x;
            Y = y;
            Z = z;
        }
        public void Move(double x, double y, double z)
        {
            X += x;
            Y += y;
            Z += z;
        }
        public Vector Cross(Vector vector) =>
            new Vector(
                Y * vector.Z - Z * vector.Y,
                Z * vector.X - X * vector.Z,
                X * vector.Y - Y * vector.X
                );
        public Vector Rotate(double angleX, double angleY, double angleZ) =>
            RotateByX(angleX).RotateByY(angleY).RotateByZ(angleZ);
        private Vector RotateByX(double angle)
        {
            angle *= Math.PI / 180;
            return new Vector(
                X,
                Y * Math.Cos(angle) + Z * Math.Sin(angle),
                -Y * Math.Sin(angle) + Z * Math.Cos(angle)
            );
        }
        private Vector RotateByY(double angle)
        {
            angle *= Math.PI / 180;
            return new Vector(
                X * Math.Cos(angle) + Z * Math.Sin(angle),
                Y,
                -X * Math.Sin(angle) + Z * Math.Cos(angle)
            );
        }
        private Vector RotateByZ(double angle)
        {
            angle *= Math.PI / 180;
            return new Vector(
                X * Math.Cos(angle) - Y * Math.Sin(angle),
                X * Math.Sin(angle) + Y * Math.Cos(angle),
                Z
            );
        }
        public static Vector operator -(Vector vectorOne, Vector vectorTwo) =>
            new Vector(
                vectorOne.X - vectorTwo.X,
                vectorOne.Y - vectorTwo.Y,
                vectorOne.Z - vectorTwo.Z
                );
    }
}

File: src/cube/geometry/Vector.java
package cube.geometry;

public class Vector {
    private double x, y, z;
    Vector(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    public double getX() {
        return x;
    }
    public double getY() {
        return y;
    }
    public double getZ() {
        return z;
    }
    void move(double x, double y, double z) {
        this.x += x;
        this.y += y;
        this.z += z;
    }
    Vector minus(Vector vector) {
        return new Vector(x - vector.x, y - vector.y, z - vector.z);
    }
    Vector cross(Vector vector) {
        return new Vector(
                y * vector.z - z * vector.y,
                z * vector.x - x * vector.z,
                x * vector.y - y * vector.x
        );
    }
    void rotate(double angleX, double angleY, double angleZ) {
        this.rotateByX(angleX);
        this.rotateByY(angleY);
        this.rotateByZ(angleZ);
    }
    private void rotateByX(double angle) {
        angle *= Math.PI / 180;
        double tempY = y * Math.cos(angle) + z * Math.sin(angle);
        z = -y * Math.sin(angle) + z * Math.cos(angle);
        y = tempY;
    }
    private void rotateByY(double angle) {
        angle *= Math.PI / 180;
        double tempX = x * Math.cos(angle) + z * Math.sin(angle);
        z = -x * Math.sin(angle) + z * Math.cos(angle);
        x = tempX;
    }
    private void rotateByZ(double angle) {
        angle *= Math.PI / 180;
        double tempX = x * Math.cos(angle) - y * Math.sin(angle);
        y = x * Math.sin(angle) + y * Math.cos(angle);
        x = tempX;
    }
}